[ENH] BEP 020 Eye Tracking (#1128)
bids-standard/bids-specification: src/modality-specific-files/physiological-recordings.md

a particular run.
For example:

<!-- This block generates a file tree.
A guide for using macros can be found at
 https://github.com/bids-standard/bids-specification/blob/master/macros_doc.md
-->
{{ MACROS___make_filetree_example({
  "sub-01": {
    "func": {
      "sub-01_task-nback_run-1_echo-1_bold.nii.gz": "",
      "sub-01_task-nback_run-1_echo-2_bold.nii.gz": "",
      "sub-01_task-nback_run-1_echo-3_bold.nii.gz": "",
      "sub-01_task-nback_run-1_physio.tsv.gz": "",
    },
  },
}) }}

This specification section first describes the organization of
[continuous physiological recordings](#continuous-physiological-recordings), and
then [events corresponding to the physiological recordings](#physiology-events).
Finally, the remainder of the document describes
[specific types of continuous recordings](#specific-physiological-signal-types)
such as [eye-tracking](#eye-tracking).

## Continuous physiological recordings

Continuous physiological recordings, such as pulse monitoring,
electrocardiogram, respiratory movement measured with a respiration belt,
gas concentration, eye-tracking, or head-motion parameters estimated

> by the MRI scanner, MUST use `_physio.<tsv.gz|json>` pairs.

Reviewer: I'm okay removing it, but head-motion is explicitly mentioned in the spec, and when I investigated about this, it seemed that head motion tracked by a camera during the scanning is out of the scope of the motion extension.

That said, I think we should try to make this BEP agnostic to that discussion. If removing is the appropriate way, I'll try to remember this conversation and bring it back after the BEP is concluded and things are settled.

corresponding tabular file but with a `.json` extension.

!!! warning "Caution"

    Columns of TSV.GZ files MUST be defined in the corresponding JSON sidecar
    and the tabular content MUST NOT include a header line.

Summary template:

<div class="highlight">
<pre><code><a href="../appendices/entities.html#sub">sub-&lt;label&gt;</a>/
    [<a href="../appendices/entities.html#ses">ses-&lt;label&gt;</a>/]
        &lt;<a href="../glossary.html#data_type-common_principles">datatype</a>&gt;/
            &lt;matches&gt;[_<a href="../appendices/entities.html#recording">recording</a>-&lt;<a href="../glossary.html#label-common_principles">label</a>&gt;]_&lt;<a href="../glossary.html#physio-suffixes">physio</a>|<a href="../glossary.html#physioevents-suffixes">physioevents</a>&gt;<a href="../glossary.html#json-extensions">.json</a>
            &lt;matches&gt;[_<a href="../appendices/entities.html#recording">recording</a>-&lt;<a href="../glossary.html#label-common_principles">label</a>&gt;]_&lt;<a href="../glossary.html#physio-suffixes">physio</a>|<a href="../glossary.html#physioevents-suffixes">physioevents</a>&gt;<a href="../glossary.html#tsvgz-extensions">.tsv.gz</a>

> </code></pre></div>

Reviewer 2: AFAIK shouldn't be HTML here -- I can't read it... Use MACRO as below or demonstrate on a simple `Text` example?

Reviewer 3: Problem is that that `matches` is not in the glossary so we can't have macro yet.

Opened this so we can have it in the glossary and then after make it easier to MACRO the :poop: out of this.
https://github.com/bids-standard/bids-specification/pull/1781

Reviewer 3: See also: https://github.com/bids-standard/bids-specification/pull/1128#discussion_r1536958111

Reviewer 3: and also... all PRs are rendered as HTML: https://bids-specification--1128.org.readthedocs.build/en/1128/modality-specific-files/physiological-recordings.html#physiology-events

😉

Reviewer: > AFAIK shouldn't be HTML here -- I can't read it... Use MACRO as below or demonstrate on a simple `Text` example?

Yes, we're using this to set some expectation for the macros to deal with this.

```{.text noheader="1" linenums="1"}
13894432329	10.1
13894432330	10.0
13894432331     9.5
13894432332     9.2
13894432333     9.0
13894432334	10.2
13894432335	10.3
13894432336	10.1
```

And `sub-01_task-nback_physio.json` defines `timestamp` column:

```JSON
{
    "SamplingFrequency": 100.0,
    "StartTime": -22.345,
    "Columns": ["timestamp", "cardiac"],
    "cardiac": {
        "Description": "continuous pulse measurement",
        "Units": "mV"
    },
    "timestamp": {
        "Description": "a continuously increasing identifier of the sampling time registered by the device",
        "Units": "ms",
        "Origin": "System startup"
    }
}
```

The decompressed contents of the corresponding `sub-01_task-nback_physioevents.tsv.gz` are:

```{.text noheader="1" linenums="1"}
13894432325	Ready
13894432331	Synchronous recalibration triggered
13894432334	External message received: new block
```

To indicate that the first column (`onset`) is to be interpreted as a timestamp,
the `ForeignIndexColumn` MUST be set to `"timestamp"` in `sub-01_task-nback_physioevents.json`:

```JSON
{
    "Columns": ["onset", "message"],
    "Description": "Messages logged by the measurement device",

>     "ForeignIndexColumn": "timestamp"
> }
> ```
> 
> If there is no appropriate source column in
> `<matches>[_recording-<label>]_physio.tsv.gz`, `ForeignIndexColumn` MAY be set to `"n/a"`.
> In this case, the values of `onset` MUST be interpreted as row indices
> into the `physio` file, with the first row having index zero (`0`).
> Negative onsets are possible, and such events are interpreted as occurring prior
> to the start of the recording, at the same sampling rate.
> 
> For example, the above `sub-01_task-nback_physioevents.tsv.gz` could be equivalently
> written:
> 
> ```{.text noheader="1"}
> -3	Ready
> 3	Synchronous recalibration triggered
> 6	External message received: new block
> ```
> 
> To indicate that the first column (`onset`) is to be interpreted as an index,
> the `ForeignIndexColumn` is set to `"n/a"` in `sub-01_task-nback_physioevents.json`:
> 
> ```JSON
> {
>     "Columns": ["onset", "message"],
>     "Description": "Messages logged by the measurement device",
>     "ForeignIndexColumn": "n/a"

Reviewer 4: ````suggestion
    "OnsetSource": "timestamp"
}
```

If there is no appropriate source column in
`<matches>[_recording-<label>]_physio.tsv.gz`, `OnsetSource` MAY be set to `"n/a"`.
In this case, the values of `onset` MUST be interpreted as row indices
into the `physio` file, with the first row having index zero (`0`).
Negative onsets are possible, and such events are interpreted as occurring prior
to the start of the recording, at the same sampling rate.

For example, the above `sub-01_task-nback_physioevents.tsv.gz` could be equivalently
written:

```{.text noheader="1"}
-3	Ready
3	Synchronous recalibration triggered
6	External message received: new block
```

To indicate that the first column (`onset`) is to be interpreted as an index,
the `OnsetSource` is set to `"n/a"` in `sub-01_task-nback_physioevents.json`:

```JSON
{
    "Columns": ["onset", "message"],
    "Description": "Messages logged by the measurement device",
    "OnsetSource": "n/a"
````

For example, the above `sub-01_task-nback_physioevents.tsv.gz` could be equivalently
written:

```{.text noheader="1"}
-3	Ready
3	Synchronous recalibration triggered
6	External message received: new block
```

To indicate that the first column (`onset`) is to be interpreted as an index,
the `ForeignIndexColumn` is set to `"n/a"` in `sub-01_task-nback_physioevents.json`:

```JSON
{
    "Columns": ["onset", "message"],
    "Description": "Messages logged by the measurement device",
    "ForeignIndexColumn": "n/a"
}
```

## Specific physiological signal types

### Eye-tracking

<!--
1. Datasets will be updated later to adapt to the agreed format.
2. We aim at adding a last example converting published dataset from openeuro.
-->
!!! example "Example datasets"

    [Example datasets](https://bids-standard.github.io/bids-examples/#dataset-index)
    with eye-tracking data have been formatted using this specification
    and can be used for practical guidance when curating a new dataset:

    -   Combined fMRI and eye-tracking data in a resting-state task, measured with
        an Eyelink (SR research). Human participant kept their gaze steady at the
        screen center.

>         [BIDS dataset](https://openneuro.org/datasets/ds004158/versions/2.0.1)

Reviewer 5: ```suggestion
        [BIDS dataset](https://openneuro.org/datasets/ds004158)
```

      "sub-01_task-visualSearch_events.tsv": "",
      "sub-01_task-visualSearch_recording-eye1_physio.json": "",
      "sub-01_task-visualSearch_recording-eye1_physio.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye1_physioevents.json": "",
      "sub-01_task-visualSearch_recording-eye1_physioevents.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye2_physio.json": "",
      "sub-01_task-visualSearch_recording-eye2_physio.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye2_physioevents.json": "",
      "sub-01_task-visualSearch_recording-eye2_physioevents.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye3_physio.json": "",
      "sub-01_task-visualSearch_recording-eye3_physio.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye3_physioevents.json": "",
      "sub-01_task-visualSearch_recording-eye3_physioevents.tsv.gz": ""
    },
  },
}) }}

Given the above example file structures, a corresponding
`sub-01_task-visualSearch_recording-eye1_physio.json` sidecar
could read:

```JSON
{
    "DeviceSerialNumber": "17535483",
    "Columns": ["timestamp", "x_coordinate", "y_coordinate", "pupil_size"],

>     "EnvironmentCoordinates": "top-left",

Reviewer 5: ```suggestion
```

replaced by `ScreenOrigin` in `events.json`

column called `"timestamp"`:

```JSON
{
    "SamplingFrequency": 100.0,
    "StartTime": -22.345,
    "Columns": ["timestamp", "cardiac"],
    "cardiac": {
        "Description": "continuous pulse measurement",
        "Units": "mV"
    },
    "timestamp": {
        "Description": "a continuously increasing identifier of the sampling time registered by the device",
        "Units": "ms",
        "Origin": "System startup"
    }
}
```

The decompressed contents of `sub-01_task-nback_physio.tsv.gz`
are now:

```{.text noheader="1" linenums="1"}
10.1	13894432329
10.0	13894432330
9.5	13894432331
9.2	13894432332
9.0	13894432333
10.2	13894432334
10.3	13894432335
10.1	13894432336
```

Then, the `sub-01_task-nback_physioevents.json` MAY define
the `ForeignIndexColumn` for indexing:

```JSON
{
    "Columns": ["onset", "message"],
    "Description": "Messages logged by the measurement device",
    "ForeignIndexColumn": "timestamp"
}
```

The decompressed contents of the corresponding TSV file,
`sub-01_task-nback_physioevents.tsv.gz` could read now:

```tsv
13894432325	Ready
13894432331	Synchronous recalibration triggered
13894432334	External message received: new block
```

> 

Reviewer 4: I'll note that the onset of the "Ready" message is in the scale of `timestamp` column, but is not a value explicitly present in that column.

That's fine, but might be worth saying explicitly that `onset` values may not correspond to exact values in the `ForeignIndexColumn`. And there will be little the validator can do on this point apart from enforce the same type in the two columns.

been acquired with an *eye-tracker*.
In the following, *eye-tracker* refers to the apparatus allowing
the recording of gaze position, and, optionally, pupil size.

Eye-tracking data MUST be stored following the general specifications
for `"generic"` physiological recordings.
However, it is REQUIRED that recordings corresponding to each eye
(and/or *cyclopean* or averaged signals for binocular eye-trackers providing
a third recording) are split into files with different
[`recording-<label>`](../appendices/entities.md#recording).
Therefore, the use of [`recording-<label>`](../appendices/entities.md#recording)
is REQUIRED with eye-tracking data.
The values `"eye1"`, `"eye2"`, and `"eye3"` are RECOMMENDED as the respective labels
for the [`recording-<label>`](../appendices/entities.md#recording) entity.

!!! danger "MANDATORY metadata"

    The correspondence of labels and the recorded eye MUST be encoded
    by the MANDATORY `RecordedEye` metadata.

    The [`recording-<label>`](../appendices/entities.md#recording) entity
    MAY take other values such as `"left"`, `"cyclopean"`, or `"right"` corresponding
    to the `RecordedEye` metadata.

>     However, it is RECOMMENDED that metadata is not encoded in the file names to avert
>     that filename and metadata enter into conflict.
>     For example, if [`recording-<label>`](../appendices/entities.md#recording) takes
>     the value `"left"` but the corresponding sidecar JSON file contains a definition of
>     `RecordedEye` being `"right"`.

Reviewer 4: If you're making a RECOMMENDATION, then presumably the validator should warn if the value is not `eye1`, `eye2` or `eye3`. You would need to write a check.

I wonder if a cleaner proposal would be:

```suggestion
    To avoid filenames and metadata coming into conflict,
    if the `recording` label starts with `left` or `right`,
    then the `RecordedEye` field MUST be `"left"` or `"right"`, respectively.
    For example `recording-left` or `recording-lefteye` MUST
    match to `"RecordedEye": "left"` in the corresponding JSON sidecar file.
```

I just suspect that you're going to get a lot of people wanting to use `left` and `right` in the filename, and erroring on mismatch will be more effective than warning on use.

Reviewer: This is the same problem of dir-. You can have `dir-AP` and `PhaseEncodingDirection` = `i` with a RAS affine and the validator can't do much about it.

What the validator can impose is that eye tracking physio recordings must have a label with pattern `/eye[1-9]`.

would that be okay?

Reviewer 4: > This is the same problem of dir-. You can have `dir-AP` and `PhaseEncodingDirection` = `i` with a RAS affine and the validator can't do much about it.

We can check this, and I'm up for making it happen.

Keeping `right`/`left` correct is also *much easier* than `AP/PA`.

> What the validator can impose is that eye tracking physio recordings must have a label with pattern `/eye[1-9]`.

Yes, this can be done. In either case, you need MUST language and an error, not a recommendation.

I'll defer to BEP020 leads to decide which will be more acceptable to users.

    "PhysioType": "eyetrack",
    "RecordedEye": "right",
    "SampleCoordinateSystem": "gaze-on-screen",
    "SamplingFrequency": 1000,
    "SoftwareVersion": "SREB1.10.1630 WIN32 LID:F2AE011 Mod:2017.04.21 15:19 CEST",
    "ScreenAOIDefinition": [
        "square",
        [100, 150, 300, 350]
    ],
    "timestamp": {
        "Description": "a continuously increasing identifier of the sampling time registered by the device",
        "Units": "ms",
        "Origin": "System startup"
    },
    "x_coordinate": {
      "LongName": "Gaze position (x)",
      "Description": "Gaze position x-coordinate of the recorded eye, in the coordinate units specified in the corresponding metadata sidecar.",
      "Units": "pixel"
    },
    "y_coordinate": {
      "LongName": "Gaze position (y)",
      "Description": "Gaze position y-coordinate of the recorded eye, in the coordinate units specified in the corresponding metadata sidecar.",
      "Units": "pixel"
    },
    "pupil_size": {
        "Description": "Pupil area of the recorded eye as calculated by the eye-tracker in arbitrary units (see EyeLink's documentation for conversion).",

>         "Units": "a.u."

Reviewer 4: We use "arbitrary" elsewhere.

```suggestion
        "Units": "arbitrary"
```

    "SamplingFrequency": 100.0,
    "StartTime": -22.345,
    "cardiac": {
      "Description": "continuous pulse measurement",
      "Units": "mV"
    },
    "respiratory": {
      "Description": "continuous measurements by respiration belt",
      "Units": "mV"
    },
    "trigger": {
      "Description": "continuous measurement of the scanner trigger signal",
      "Units": "V"
    }
}
```

The example shows the three REQUIRED metadata entries `Columns`, `SamplingFrequency`,
and `StartTime`.
Columns are further described following the specifications for
[tabular files](../common-principles.md#tabular-files),
indicating `Description` and `Units` fields.
Other fields, such as `TermURL`, `LongName`, MAY be included.

The `PhysioType` can be omitted as the recordings are `"generic"`:

```JSON
{
    "Columns": ["cardiac", "respiratory", "trigger"],
    "Manufacturer": "Brain Research Equipment ltd.",
    "SamplingFrequency": 100.0,
    "StartTime": -22.345,
    "cardiac": {
      "Description": "continuous pulse measurement",
      "Units": "mV"
    },
    "respiratory": {
      "Description": "continuous measurements by respiration belt",
      "Units": "mV"
    },
    "trigger": {

>       "Description": "continuous measurement of the scanner trigger signal"

Reviewer 3: TODO: add unit

0.04 s before the first sample, therefore at -22.305 s.

**Explicit indexing of `<matches>_physioevents.tsv.gz` files (RECOMMENDED)**.
When the `ForeignIndexColumn` defines a value such as `"timestamp"`,
that specific column name MUST be present the `<matches>_physio.tsv.gz` file.
In that case, all values of the `"onset"` column of the
`<matches>_physioevents.tsv.gz` are indexed by the index established by
the `ForeignIndexColumn` of the corresponding `<matches>_physio.tsv.gz` file
(that is, `"timestamp"` in this example).

For example, let us consider the previous structure:

<!-- This block generates a file tree.
A guide for using macros can be found at
 https://github.com/bids-standard/bids-specification/blob/master/macros_doc.md
-->
{{ MACROS___make_filetree_example({
  "sub-01": {
    "func": {
      "sub-01_task-nback_physio.json": "",
      "sub-01_task-nback_physio.tsv.gz": "",
      "sub-01_task-nback_physioevents.json": "",
      "sub-01_task-nback_physioevents.tsv.gz": "",
    },
  },
}) }}

However, the `sub-01_task-nback_physio.json` will define now an extra
column called `"timestamp"`:

```JSON
{
    "SamplingFrequency": 100.0,
    "StartTime": -22.345,

>     "Columns": ["cardiac", "timestamp"],

Reviewer 3: maybe better if timestamp comes first, no?

      "sub-01_task-visualSearch_bold.nii.gz": "",
      "sub-01_task-visualSearch_events.json": "",
      "sub-01_task-visualSearch_events.tsv": "",
      "sub-01_task-visualSearch_recording-eye1_physio.json": "",
      "sub-01_task-visualSearch_recording-eye1_physio.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye1_physioevents.json": "",
      "sub-01_task-visualSearch_recording-eye1_physioevents.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye2_physio.json": "",
      "sub-01_task-visualSearch_recording-eye2_physio.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye2_physioevents.json": "",
      "sub-01_task-visualSearch_recording-eye2_physioevents.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye3_physio.json": "",
      "sub-01_task-visualSearch_recording-eye3_physio.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye3_physioevents.json": "",
      "sub-01_task-visualSearch_recording-eye3_physioevents.tsv.gz": ""
    },
  },
}) }}

Given the above example file structures, a corresponding
`sub-01_task-visualSearch_recording-eye1_physio.json` sidecar
could read:

```JSON
{
    "DeviceSerialNumber": "17535483",

>     "Columns": ["timestamp", "x_coordinate", "y_coordinate", "pupil_size"],

Reviewer 5: In the src/schema/objects/columns.yaml and src/schema/rules/tabular_data/physio.yaml it says "et_timestamp", here it says "timestamp", which is correct?

Reviewer: Both are correct, these are two different things:

- `src/schema/objects/columns.yaml`, `src/schema/rules/tabular_data/physio.yaml` define a `et_timestamp` column name. This is so to allow the timestamp be different from the already existing ``timestamp`` column. This identifier is internal to the schema, so it does not propagate into actual column names.

- the line 672 you commented on is just the necessary specification of the `"Columns"` metadata, the column name is arbitrary, and in this case it is named `"timestamp"` but other names would be equally supported.

    "EnvironmentCoordinates": "top-left",
    "Manufacturer": "SR-Research",
    "ManufacturersModelName": "EYELINK II CL v4.56 Aug 18 2010",
    "RecordedEye": "right",
    "SampleCoordinateSystem": "gaze-on-screen",
    "SampleCoordinateUnits": "pixel",
    "SamplingFrequency": 1000,
    "SoftwareVersion": "SREB1.10.1630 WIN32 LID:F2AE011 Mod:2017.04.21 15:19 CEST",
    "ScreenAOIDefinition": [
        "square",
        [100, 150, 300, 350]
    ],
    "pupil_size": {
        "Description": "Pupil area of the recorded eye as calculated by the eye-tracker in arbitrary units (see EyeLink's documentation for conversion).",

>         "Units": "a.u."

Reviewer 3: this only describes pupil_size but the tsv content below has 4 columns, wouldn't it be better to have all 4 described?

I think that users can figure it out but I would prefer to avoid having users interprate the specification and be explicit about it.

Reviewer 5: I put all descriptions to be written in the json file by the converter...I thought this was just an example to show that all descriptions should be present

Reviewer: @Remi-Gau c0401594

    }
}
```

> Content of `sub-01_task-VisualSearch_events.json`:

Reviewer 5: Is this really the _events.json file or is it the _physioevents.json?

Reviewer: This is correct, it annotates the events.tsv file, as opposed to the new _physioevents.json.

Reviewer 5: So basically in the _physioevents.json file I only have the column names plus their description? And do I need a _events.json file if I do not have an _events.tsv? Because Martins dataset on Open Neuro does not have it.

Reviewer: Okay, gotcha -- you've actually hit an edge case we do not currently cover:

- We say that some metadata belonging in the `_events.json` file becomes mandatory if eye tracking is present
- However, BIDS allows `task-rest` to dismiss the events.tsv file. This BEP could make it mandatory for rest, even if the `events.tsv` file is just empty (only the header row).

WDYT @effigies @Remi-Gau ?

Reviewer: (dataset: https://doi.org/10.18112/openneuro.ds004158.v2.0.2)

Reviewer 5: Or include it in the `physioevents.json`? (if this does not disturb too much other stuff) Because that would be easy to include in the converter as we are already asking users to put some metadata manually in the` metadata.yml ` which is read in by our code. The current code actually does that because I understood it wrong, cf my initial question...

Reviewer: Not sure that is a good idea. If I were to move the metadata from events.json, I would think it'd be better in the physio.json file rather than physioevents.json.

These metadata are about stimulus presentation, so they are likely relevant beyond eye tracking exclusively. Perhaps a reasonable alternative would be to update the specs so these metadata are encoded within stim.json files instead of the events.json files. Either way, they need to allow a json be defined without the corresponding .tsv[.gz] file, or allow empty tables (okay for .tsv but dangerous for .tsv.gz).

Reviewer 5: Ok, I'm of course fine with the solution that fits best to BIDS in general but I agree that we should keep this information somehow. 

> allow a json be defined without the corresponding .tsv[.gz] file

would probably the best solution for this particular BEP as we agreed on .tsv.gz

Reviewer 5: I agree but I think I cannot resolve conversations here

7186806    72     fixation    0      n/a
7186879    231    saccade     1      n/a
7187111    6186   fixation    0      n/a
7193298    216    saccade     1      n/a
7193515    1286   fixation    0      n/a
7194802    24     saccade     0      n/a
7194827    2403   fixation    0      n/a
7197231    17     saccade     0      n/a
7197249    1640   fixation    0      n/a
7198890    6      saccade     0      n/a
7198897    1105   fixation    0      n/a
7200003    233    saccade     1      n/a
7200237    184    fixation    0      n/a
7200422    15     saccade     0      n/a
7200438    264    fixation    0      n/a
```

where the first three rows are logged by the eye-tracker and the fourth row shows
a *message* asynchronously received by the eye-tracker.
The remainder of the rows contain a subset of parameters registered by the device,
derived from applying a model to identify saccades and blinks.
The corresponding `sub-01_task-visualSearch_recording-eye1_physioevents.json` sidecar
would read:

```JSON
{
    "Columns": ["onset", "duration", "trial_type", "blink", "message"],
    "Description": "Messages logged by the measurement device",
    "ForeignIndexColumn": "timestamp",

>     "blink": {

Reviewer 2: shameless plug: this is the situation I would like to avoid in
- https://github.com/bids-standard/bids-2-devel/issues/71

where I guess such file would look like

```json
{
    "Description": "Messages logged by the measurement device",
    "ForeignIndexColumn": "timestamp",
    "Columns": {
       "onset": { "Index": 1},
       "duration": {"Index": 2},
       "trial_type": {
           "Description": "Event type as identified by the eye-tracker's model (either 'n/a' if not applicabble, 'fixation', or 'saccade')."
           "Index": 3}, 
       "blink": {
         "Description": "One indicates if the eye was closed, zero if open."
         "Index" : 4},
       "message": {
         "Description": "String messages logged by the eye-tracker.",
         "Index": 5}
}
```

Reviewer: I'm +1 on this and the solution.

However, I understand that can only be implemented in BIDS 2.0 because would introduce backward breaking changes, right?

If so, I think this is something we cannot resolve here :(

Reviewer: I'm marking resolved and defer the conversation to bids-standard/bids-2-devel#71.

>       "Description": "One indicates if the eye was closed, zero if open."

Reviewer 3: TODO Description of the levels should be done in Levels array

    "func": {
      "sub-01_task-visualSearch_bold.json": "",
      "sub-01_task-visualSearch_bold.nii.gz": "",
      "sub-01_task-visualSearch_events.json": "",
      "sub-01_task-visualSearch_events.tsv": "",
      "sub-01_task-visualSearch_recording-eye1_physio.json": "",
      "sub-01_task-visualSearch_recording-eye1_physio.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye1_physioevents.json": "",
      "sub-01_task-visualSearch_recording-eye1_physioevents.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye2_physio.json": "",
      "sub-01_task-visualSearch_recording-eye2_physio.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye2_physioevents.json": "",
      "sub-01_task-visualSearch_recording-eye2_physioevents.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye3_physio.json": "",
      "sub-01_task-visualSearch_recording-eye3_physio.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye3_physioevents.json": "",
      "sub-01_task-visualSearch_recording-eye3_physioevents.tsv.gz": ""
    },
  },
}) }}

Given the above example file structures, a corresponding
`sub-01_task-visualSearch_recording-eye1_physio.json` sidecar
could read:

> ```JSON

Reviewer 3: insert in example here

Reviewer: What example is missing here?

Reviewer: bumping up... what's missing here?

Reviewer: @Remi-Gau can you remember the mising example? Can this be closed?

      "sub-01_task-visualSearch_recording-eye1_physio.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye1_physioevents.json": "",
      "sub-01_task-visualSearch_recording-eye1_physioevents.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye2_physio.json": "",
      "sub-01_task-visualSearch_recording-eye2_physio.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye2_physioevents.json": "",
      "sub-01_task-visualSearch_recording-eye2_physioevents.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye3_physio.json": "",
      "sub-01_task-visualSearch_recording-eye3_physio.tsv.gz": "",
      "sub-01_task-visualSearch_recording-eye3_physioevents.json": "",
      "sub-01_task-visualSearch_recording-eye3_physioevents.tsv.gz": ""
    },
  },
}) }}

Given the above example file structures, a corresponding
`sub-01_task-visualSearch_recording-eye1_physio.json` sidecar
could read:

```JSON
{
    "DeviceSerialNumber": "17535483",
    "Columns": ["timestamp", "x_coordinate", "y_coordinate", "pupil_size"],
    "EnvironmentCoordinates": "top-left",
    "Manufacturer": "SR-Research",
    "ManufacturersModelName": "EYELINK II CL v4.56 Aug 18 2010",

>     "RecordedEye": "right",

Reviewer: Did we forget this?? 👀 

```suggestion
    "PhysioType": "eyetrack",
    "RecordedEye": "right",
```

Reviewer 5: It's described in the text right after the Eye tracking section starts but it wasn't in the example so far. I actually mentioned this a couple of meetings ago because I overlooked this part in the text and therefore didn't have it in the converter (it is now). So yes, we should show it in the example too 😄

    "SampleCoordinateSystem": "gaze-on-screen",
    "SampleCoordinateUnits": "pixel",
    "SamplingFrequency": 1000,
    "SoftwareVersion": "SREB1.10.1630 WIN32 LID:F2AE011 Mod:2017.04.21 15:19 CEST",
    "ScreenAOIDefinition": [
        "square",
        [100, 150, 300, 350]
    ],
    "pupil_size": {
        "Description": "Pupil area of the recorded eye as calculated by the eye-tracker in arbitrary units (see EyeLink's documentation for conversion).",
        "Units": "a.u."
    },
    "timestamp": {
        "Description": "a continuously increasing identifier of the sampling time registered by the device",
        "Units": "ms",
        "Origin": "System startup"
    }
}
```

Content of `sub-01_task-VisualSearch_events.json`:

```JSON
{
   "TaskName": "Visual Search",
   "InstitutionName": "Stanford University",
   "InstitutionAddress": "450 Serra Mall, Stanford, CA 94305-2004, USA",
   "StimulusPresentation": {

>        "ScreenDistance": 60,

Reviewer 6: ```suggestion
       "ScreenDistance": .6,
```

According to the [draft spec](https://bids-specification--1128.org.readthedocs.build/en/1128/modality-specific-files/physiological-recordings.html#eye-tracking), `ScreenDistance` should be in meters right? :

_"Distance (in meters) between the eye-tracker and the participant eye(s). "_

    -   Combined behavior and eye-tracking fixation and saccade data,
        measured with an Eyelink (SR Research), from 8 participants reading 320
        embedded target words and invisible boundary (see
        [Gagl, 2016](https://peerj.com/articles/2467/)).

        [BIDS dataset](https://zenodo.org/records/1228659)

    -   Combined behavior and eye-tracking position and pupil data, measured with
        an Eyelink (SR Research), from 26 participants performing a
        binocular rivalry task (see
        [Brascamp et.al, 2021](https://doi.org/10.7554/eLife.66161)).

        [BIDS dataset](https://doi.org/10.5061/dryad.41ns1rncp)

    -   Combined resting-state fMRI and eye-tracking data, measured with an Eyelink
        (SR research), from 20 participants keeping their gaze steady at the
        screen center.

        [BIDS dataset](https://openneuro.org/datasets/ds004158/versions/2.0.1)

Setting `PhysioType` to the keyword `"eyetrack"` specifies that
the physiological recordings in the `<matches>_physio.tsv.gz` have
been acquired with an *eye-tracker*.
In the following, *eye-tracker* refers to the apparatus allowing

> the recording of gaze position and/or pupil size.

Reviewer 6: ```suggestion
the recording of gaze position, and, optionally, pupil size.
```

The `physio.PhysioEyeTracking` columns table that appears below specifies that gaze data are REQUIRED but pupil data are OPTIONAL.  So If I understand the sentence above, "_gaze position and/or pupil size"_ is incorrect because pupil size is not allowed on its own.

Alternatively, we could change the spec to allow pupil size data without eyegaze data (no strong feelings here!)
